### DISCUSSION TITLE
Shortening up OBX segments

### DISCUSSION THREAD

**Discussion by dmyers-afk** [nextgenhealthcare/connect#6235] [Q&A] [answered]
So I'm trying to clean up the OBX segments in what I'm being sent by our reports software. Unfortunately it often sends whole paragraphs in a single OBX segment, which when received by the receiver, it mimics the formatting, which is a problem.
I found the following post:
https://github.com/nextgenhealthcare/connect/discussions/6162
BUT ... not getting it to work. Probably a me problem. So have some questions specifically about the suggested solution:

````JS
// All credit to pacmano1 for this code btw - shown in the post but wanted to make that clear.
function splitIntoChunks(str, chunkSize) {
    const chunks = [];
    var i = 0;
    while (i < str.length) {
        var end = Math.min(i + chunkSize, str.length);
        if (end < str.length && str[end] !== ' ') {
            // Find the last space in the substring
            const lastSpace = str.lastIndexOf(' ', end);
            if (lastSpace > i) {
                end = lastSpace;
            }
        }
        chunks.push(str.slice(i, end));
        i = end + (str[end] === ' ' ? 1 : 0); // Skip the space if it's the end character
    }
    return chunks;
}

var sometext = ''
for each (obx in msg.OBX) {
    sometext += obx['OBX.5'][0]['OBX.5.1'].toString() + ' '; // Ensure spaces between concatenated parts
}
msg.OBX = [] // clear current OBXs

var split = splitIntoChunks(sometext.trim(), 35)
var new_obx;
for (var i = 0; i < split.length; i++) {
    if (i == 0) {
        new_obx = createSegmentAfter('OBX', msg.OBR);
    } else {
        new_obx = createSegmentAfter('OBX', new_obx);    
    }
    new_obx['OBX.1']['OBX.1.1'] = i + 1;
    new_obx['OBX.5']['OBX.5.1'] = split[i];  
}
````

So an example section of an HL7 message being processed looks like this:
````
OBR|3|WW4593556|WW4593556|||||||||||202406270934||AAA^Last^First^||||||202406260739||F|
OBX|1||MRI : Brain w/wo|4|						
OBX|1||MRI : Brain w/wo|5|RADIOLOGY REPORT
OBX|1||MRI : Brain w/wo|6|
OBX|1||MRI : Brain w/wo|7|NAME:  LAST, FIRST 	DATE OF EXAM: 06-25-2024 
OBX|1||MRI : Brain w/wo|8|
OBX|1||MRI : Brain w/wo|9|DATE OF BIRTH:  01-01-2025 	CHART #: 999999-WW 
OBX|1||MRI : Brain w/wo|10|
OBX|1||MRI : Brain w/wo|11|PHYSICIAN:  First Last, MD 
OBX|1||MRI : Brain w/wo|12|
OBX|1||MRI : Brain w/wo|13|MRI OF THE BRAIN WITH AND WITHOUT CONTRAST 
OBX|1||MRI : Brain w/wo|14|
OBX|1||MRI : Brain w/wo|15|CLINICAL INDICATION:   Headaches, migraines.  
OBX|1||MRI : Brain w/wo|16|
OBX|1||MRI : Brain w/wo|17|Multiplanar and multiparametric MRI imaging of the brain was performed.
OBX|1||MRI : Brain w/wo|18|
OBX|1||MRI : Brain w/wo|19|ADC map and diffusion weighted acquisition reveal no area of restriction to suggest acute infarction.   
OBX|1||MRI : Brain w/wo|20|
OBX|1||MRI : Brain w/wo|21|Sagittal T1 shows no cerebellar tonsillar herniation.  Pontine and midbrain regions are unremarkable.    Optic chiasm and pituitary gland signal were unremarkable.  
OBX|1||MRI : Brain w/wo|22|
OBX|1||MRI : Brain w/wo|23|Axial T1 shows no conspicuous T1 shortening region.   
OBX|1||MRI : Brain w/wo|24|
OBX|1||MRI : Brain w/wo|25|Axial T2 shows sellar and suprasellar regions to appear unremarkable.    The 7th and 8th nerves were symmetric bilaterally.   Cerebellar pontine angle presentation was unremarkable.   No evidence of any cerebellar pontine angle mass.   The mastoid air cell signal presentation unremarkable.  Normal flow void of basilar, carotid and middle cerebral distribution was identified.   No conspicuous paraventricular T2 prolongation to suggest demyelinating or dysmyelinating process or small vessel ischemic change identified.   
OBX|1||MRI : Brain w/wo|26|
OBX|1||MRI : Brain w/wo|27|Paranasal sinuses appear patent.   Normal gray white matter signal differentiation.   Globes, optic nerves and orbital muscles and periorbital muscles are symmetric.  Retroorbital fat signal unremarkable.   
OBX|1||MRI : Brain w/wo|28|
````

So to get more into my questions:

````JS
var sometext = ''
for each (obx in msg.OBX) {
    sometext += obx['OBX.5'][0]['OBX.5.1'].toString() + ' '; // Ensure spaces between concatenated parts
}
msg.OBX = [] // clear current OBXs
````
So this cycles through all of the OBX segments, taking OBX segment 5, the actual data and adding it into the variable sometext. Effectively a super long string of all the text in the OBX segment.
Once that is done it deletes all OBX segments.

So at this point , in theory the message goes to the OBR, and all of the OBX segments are gone.

````JS
var split = splitIntoChunks(sometext.trim(), 35)

function splitIntoChunks(str, chunkSize) {
    const chunks = [];
    var i = 0;
    while (i < str.length) {
        var end = Math.min(i + chunkSize, str.length);
        if (end < str.length && str[end] !== ' ') {
            // Find the last space in the substring
            const lastSpace = str.lastIndexOf(' ', end);
            if (lastSpace > i) {
                end = lastSpace;
            }
        }
        chunks.push(str.slice(i, end));
        i = end + (str[end] === ' ' ? 1 : 0); // Skip the space if it's the end character
    }
    return chunks;
}
````
So this part, as far as I am following is creating an array of text from "sometext". Basically it's getting the first 35 characters, checking if there is a space, then returning the array.
So now we have "split", which is an array of (up to) 35 character strings from "sometext"

````JS
var new_obx;
for (var i = 0; i < split.length; i++) {
    if (i == 0) {
        new_obx = createSegmentAfter('OBX', msg.OBR);
    } else {
        new_obx = createSegmentAfter('OBX', new_obx);    
    }
    new_obx['OBX.1']['OBX.1.1'] = i + 1;
    new_obx['OBX.5']['OBX.5.1'] = split[i];  
}
````
Finally, it cycles through the array "split", creates an initial OBX segment after the OBR segment, then subsequent OBX segments after that, renumbering them as it goes.

All good stuff. But I'm obviously missing something as that part of the transformer doesn't appear to be doing anything to the message.

That said, I had a couple of thoughts:
1. I'm noticing that the HL7 message is starting with OBX|1||MRI : Brain w/wo|4| where I would probably have expected OBX|1||MRI : Brain w/wo|1|. That's not a typo, so maybe that's throwing things off. I haven't found a good link on exactly how
````JS
for each (obx in msg.OBX)
````
works, so maybe that strange starting count throws it off?
2. Maybe I'm supposed to have something external to the transfomer, or a mapping or something that I'm just missing?
3. Is there a way to call an external program? I'm not at all familiar with JavaScript, but am fairly proficient in a few others. Wasn't sure if I could call a different program to clean up the text or if I need to keep plowing through learning JavaScript on the fly.
4. I've tried using the debugger in Mirth, but it's on a VM and I'm getting a "headless java" error, which apparently is due to it not seeing a display and/or keyboard.

Thank you for your patience. I am trying to learn.

**Comment by pacmano1** [comment]
Read the user manual starting at https://docs.nextgen.com/bundle/Mirth_User_Guide_40/page/connect/connect/topics/c_mirth_connect_debugger_connect.html.   I think you missed a few things.

That's a long post (and who knows if my code works, lol).   Post a before and after message please.  It's often faster (for me) to just see what you are expecting.

**Reply by dmyers-afk** [reply]
I'll read up on the debugger.

TLDR is I have OBX segments like this:
````
OBX|1||MRI : Brain w/wo|25|Axial T2 shows sellar and suprasellar regions to appear unremarkable.    The 7th and 8th nerves were symmetric bilaterally.   Cerebellar pontine angle presentation was unremarkable.   No evidence of any cerebellar pontine angle mass.   The mastoid air cell signal presentation unremarkable.  Normal flow void of basilar, carotid and middle cerebral distribution was identified.   No conspicuous paraventricular T2 prolongation to suggest demyelinating or dysmyelinating process or small vessel ischemic change identified.   
````
That I want to be able to get into more like this:
````
OBX|25||MRI : Brain w/wo|25|Axial T2 shows sellar and suprasellar regions to appear unremarkable.    The 
OBX|26||MRI : Brain w/wo|26|7th and 8th nerves were symmetric bilaterally.   Cerebellar pontine angle 
OBX|27||MRI : Brain w/wo|27|presentation was unremarkable.   No evidence of any cerebellar pontine angle 
OBX|28||MRI : Brain w/wo|28|mass.   The mastoid air cell signal presentation unremarkable.  Normal flow 
OBX|29||MRI : Brain w/wo|29|void of basilar, carotid and middle cerebral distribution was identified.   No 
OBX|29||MRI : Brain w/wo|29|conspicuous paraventricular T2 prolongation to suggest demyelinating or 
OBX|30||MRI : Brain w/wo|30|dysmyelinating process or small vessel ischemic change identified.   
````
So just being able to take OBX 5 and make it into shorter lines (ex. only 80 characters long) based on spaces etc.

Thanks

**Reply by dmyers-afk** [reply]
I'm definitely missing something. I added a prior transformer with something simple:
````JS
var i=0;
for each (OBX in msg.OBX) {
    msg['OBX'][i]['OBX.4'] = i;
    msg['OBX'][i]['OBX.1] = i;
    i++;
}
````
Thinking that would renumber my OBX segments. That also makes no changes to the OBX segments. So perhaps is it the numbering of the OBX segments? I would think the iteration would work based on the start of the segment, and not be concerned about the number, but I could be wrong.

**Comment by pacmano1** [comment, marked as answer]
I would not use my chunker function anymore.  @tonygermano pointed out that Apache WordUtils is a better idea.

Not fully QAed, but:

```JS
var WordUtils = Packages.org.apache.commons.lang3.text.WordUtils;
var largeString = msg['OBX']['OBX.5']['OBX.5.1'].toString()
var obx_id = msg['OBX']['OBX.3']['OBX.3.1'].toString()
delete msg.OBX
var maxLength = 30;
var wrappedString = WordUtils.wrap(largeString, maxLength);
var brokenStringsArray = wrappedString.split("\n");
// Output the result
var new_obx;
for (var i = 0; i < brokenStringsArray.length; i++) {
    if (i == 0) {
        new_obx = createSegmentAfter('OBX', msg.OBR);
    } else {
        new_obx = createSegmentAfter('OBX', new_obx);    
    }
    new_obx['OBX.1']['OBX.1.1'] = i + 1;
    new_obx['OBX.3']['OBX.3.1'] = obx_id
    new_obx['OBX.4']['OBX.4.1'] = new_obx['OBX.1']['OBX.1.1']
    new_obx['OBX.5']['OBX.5.1'] = brokenStringsArray[i];  
}
```

**Reply by dmyers-afk** [reply]
If I'm following, this is just doing a single OBX line correct? So I need to wrap this in a for each loop to go through each OBX segment?

**Reply by pacmano1** [reply]
Yes, it takes one OBX and word wraps that OBX.   Are you saying you have many long lines of OBXs that you need to wrap?  If yes, then your before and after posting didn't reflect that.

**Reply by dmyers-afk** [reply]
I do. Sorry about that. Was trying to condense the discussion down and wasn't specific enough. I just wanted to make sure I wasn't missing something obvious in the prior code. Thanks

**Reply by dmyers-afk** [reply]
So just to try out the code I took a sample HL7 message and ran it through mirth where it has the regular things and one OBX segment with a super long bit of text. It doesn't appear to make any changes to the OBX line at all. I made sure the OBX numbering was 1 and same thing. So I'm back at square one why it doesn't appear to edit the OBX line at all

**Reply by pacmano1** [reply]
Post a _full_ message.   All of it.  MSH, PID , etc.  Just anonymize it.

**Reply by dmyers-afk** [reply]
So here's an example with a single OBX line. 
````
MSH|^~\&|ISERIES|01|RAD|RAD|20240628141909||ORU^O01|01127473|P|2.3
PID|1|123456-WW|123456-WW|123456-WW|LAST^FIRST^||12311900|M|||||
PV1|2||||||1565^Last^First^|1565^Last^First^|CT Scan : Chest
OBR|3|WW4188874|WW4188874|||||||||||202406271308||1565^Last^First^||||||202406260858||F|
OBX|1||CT Scan : Chest|1|Lung windows exhibit right apical paraseptal emphysematous lung change that appears mild.   Generalized mild to moderate hyperinflation noted with areas of bibasilar atelectasis versus scar.   No focal alveolar consolidation.    No pleural effusion or pneumothorax.    Scattered granulomata identified.
````
A longer example would be:
````
MSH|^~\&|ISERIES|01|RAD|RAD|20240628141909||ORU^O01|01127473|P|2.3
PID|1|123456-WW|123456-WW|123456-WW|LAST^FIRST^||12311900|M|||||
PV1|2||||||1565^Last^First^|1565^Last^First^|CT Scan : Chest
OBR|3|WW4188874|WW4188874|||||||||||202406271308||1565^Last^First^||||||202406260858||F|OBX|1||CT Scan : Chest|4|				
OBX|1||CT Scan : Chest|5|RADIOLOGY REPORT
OBX|1||CT Scan : Chest|6|
OBX|1||CT Scan : Chest|7|NAME:  LAST, FIRST 	DATE OF EXAM: 06-26-2024 
OBX|1||CT Scan : Chest|8|
OBX|1||CT Scan : Chest|9|DATE OF BIRTH:  12-31-1900 	CHART #: 123456-WW 
OBX|1||CT Scan : Chest|10|
OBX|1||CT Scan : Chest|11|PHYSICIAN:  Person, MD 
OBX|1||CT Scan : Chest|12|
OBX|1||CT Scan : Chest|13|CT SCAN OF THE CHEST - LUNG CANCER SCREENING
OBX|1||CT Scan : Chest|14|
OBX|1||CT Scan : Chest|15|CLINICAL INDICATION:   Smoking greater than 40 pack years.
OBX|1||CT Scan : Chest|16|
OBX|1||CT Scan : Chest|17|Multiplanar CT review of the chest was performed without contrast.  
OBX|1||CT Scan : Chest|18|
OBX|1||CT Scan : Chest|19|Scout image shows sternotomy apparatus. 
OBX|1||CT Scan : Chest|20|
OBX|1||CT Scan : Chest|21|Lung windows exhibit right apical paraseptal emphysematous lung change that appears mild.   Generalized mild to moderate hyperinflation noted with areas of bibasilar atelectasis versus scar.   No focal alveolar consolidation.    No pleural effusion or pneumothorax.    Scattered granulomata identified.   
OBX|1||CT Scan : Chest|22|
OBX|1||CT Scan : Chest|23|Soft tissues show normal attenuation of thyroid gland.   No supraclavicular, axillary or mediastinal adenopathy identified.    Right upper extremity central venous catheter identification with tip terminating at mid to low SVC level.   
OBX|1||CT Scan : Chest|24|
OBX|1||CT Scan : Chest|25|Aortic atherosclerosis.  Normal arch anatomy.  Great vessel atherosclerosis.   No aortic aneurysm.  Diffuse coronary atherosclerosis.   Valvular calcification.   No pericardial effusion.   Small hiatal hernia.   
OBX|1||CT Scan : Chest|26|
OBX|1||CT Scan : Chest|27|Upper abdomen showed incidental left renal cyst presentation. 
OBX|1||CT Scan : Chest|28|
OBX|1||CT Scan : Chest|29|
OBX|1||CT Scan : Chest|30|
OBX|1||CT Scan : Chest|31|No compression fracture or subluxation.  Sternotomy apparatus.   No scoliosis.   
OBX|1||CT Scan : Chest|32|
OBX|1||CT Scan : Chest|33|Visceral lymph node versus subpleural ground glass nodule appreciation involving right major fissure at the level of the distal right main stem bronchus is appreciated measuring approximately 4.0 mm.    
OBX|1||CT Scan : Chest|34|
OBX|1||CT Scan : Chest|35|IMPRESSION:
OBX|1||CT Scan : Chest|36|1. Hyperinflation with multifocal areas of scarring. 
OBX|1||CT Scan : Chest|37|2. Scattered calcified lung nodularity.
OBX|1||CT Scan : Chest|38|3. Ground glass nodule right lower lobe abutting major fissure in subpleural fashion versus intra-fissure lymph node as noted.   Six month interval surveillance program advised as warranted. 
OBX|1||CT Scan : Chest|39|4. Other findings as noted above.  
OBX|1||CT Scan : Chest|40|
OBX|1||CT Scan : Chest|41|
````
Thanks

**Reply by pacmano1** [reply]
![image](https://github.com/nextgenhealthcare/connect/assets/44065187/9ec0f7e8-fa31-45dd-bfd5-a8abd28b89b1)

In message (same message as above)

```
MSH|^~\&|ISERIES|01|RAD|RAD|20240628141909||ORU^O01|01127473|P|2.3
PID|1|123456-WW|123456-WW|123456-WW|LAST^FIRST^||12311900|M|||||
PV1|2||||||1565^Last^First^|1565^Last^First^|CT Scan : Chest
OBR|3|WW4188874|WW4188874|||||||||||202406271308||1565^Last^First^||||||202406260858||F|
OBX|1||CT Scan : Chest|1|Lung windows exhibit right apical paraseptal emphysematous lung change that appears mild.   Generalized mild to moderate hyperinflation noted with areas of bibasilar atelectasis versus scar.   No focal alveolar consolidation.    No pleural effusion or pneumothorax.    Scattered granulomata identified.
```

And the result:

![image](https://github.com/nextgenhealthcare/connect/assets/44065187/07ca62b3-1b14-4e21-b641-7844c1b559fd)

**Reply by dmyers-afk** [reply]
So this is what I'm getting:
![Transformer](https://github.com/nextgenhealthcare/connect/assets/173870822/90b55b2e-7f2c-4686-ab8a-04798011cb8d)
![Channel Msg Dest](https://github.com/nextgenhealthcare/connect/assets/173870822/beb74c97-2dd1-445b-b271-292da296b84e)
![Channel Msg Source](https://github.com/nextgenhealthcare/connect/assets/173870822/18c9939f-b8a7-4104-bea9-2cbfd4118607)

**Reply by pacmano1** [reply]
Create a brand-new channel and apply the code I posted and see if you get the same results.  You should have no filters, no message templates, just a single transformer with the code.  You can just make it a channel reader and channel writer and use the message viewer (or menu if on the dashboard) to send a message in.

**Reply by dmyers-afk** [reply]
In a new channel it works as expected (with additional new lines)
![image](https://github.com/nextgenhealthcare/connect/assets/173870822/381397a7-cc70-4829-b402-6066fc539830)
Not really sure why it being the only thing vs later on matters, but I can probably work with that. I'm going to move it to first priority in my other channel and see if that makes a difference.

**Reply by dmyers-afk** [reply]
Moving it from 7 to 1 and it works as expected. Can you not have multiple JavaScript transformers?

**Reply by pacmano1** [reply]
Yes, you can have lots of transformers.   You are probably doing something with msg in another step.   

I am a bit past the point of helping here.   Hopefully someone else will chime in.

You can also engage a consultant (I am one and work at https://diridium.com) and there are many out there.

**Reply by dmyers-afk** [reply]
You've been a HUGE help. Thank you
